notebooks/Numba_052_refpruner.ipynb

```diff
@@ -0,0 +1,656 @@
ACcCgqBv4JuclDlu3A7C/eD3J1t0EW8IAE5gBhFrCCLdEkWs3fZlz7WnYSgM2zzQBnO7/d9obsjgBo2CsmVAwJtKwYmNNk4qaKqW5WJe/hGJnAf/LfYsCRzbnRl5J57BzeOT45cw5cT2HrTvoGX4hYZYIBkVdzwtpt7KvaB3BllBGxU3nXyeuY37Gn4gOasUp+1Ey7SmV0txVtIp8RyaJBd+QY09le+FyDlm3kV5yJNDF9ys1UzYTycnJjBs3LtDZEGmyXDfH5OXl0aZNmwDnRqThDYuexLDoSRWWzYhfRK79FDHBrbC41ff8MX4hVmdhWf+xKa3+wU0tHybMEondWUKwJcRrmv0jRvJOj6yy57M6fN6QRQoYBWPNgMPhICUlhc6dm9atwCKnE/dgTKQ5c3W291Tekb/ic1cg1pypmbIZSE1NxWq1qplSpAEpGBORulIw1gz8f/buOz6KOv8f+Gtbek9IgSQkJCGQAArSixFR8ACPKO0ETuQUFZVDwymeemf5qhz+RPDUs5xyHkgJIKLA0USK9IQWUihJSO9l07N1fn+ErCm7aSSZ7Ob1fDx4PNjN7MxrZ2cn73w+n/lMVlYWALAYI+pC9WPGeEUlEbUXi7FeICsrCxKJhN2URF3IyckJUqkUSqWy9YWJiBpgMdYLZGdnw8PDAzY2NmJHIbJYMpkMzs7OKCkpETsKEZkZFmO9QHZ2NlvFiLqBm5sbiouLxY5BRGaGlzD0ApxjjKh7uLu7s2WMusV9/7ITOwJ1IhZjvUBWVhZCQ0PFjkFk8dzc3FiMUbfIuHgRZbm5Ysfo1Zx9fOA/YkSnrIvFWC+QlZWF+++/X+wYRBaP3ZTUXcpyc1Fw86bYMaiTcMxYL5CTk8NuSqJuwG5KIuoIFmMWrqioCFVVVSzGiLoBW8aIqCNYjFm45ORkAEBwcLDISYgsH8eMkZhUGk2Lj7uKTq/vketqj4b7qrv2W0MsxixccnIyrK2t4efnJ3YUIovn7u6O4uJiCIIgdhTqRRLT0rBk9WocOHfO6OOutuLjj3HiypUet662aLivunu/NcQB/BYuOTkZgYGBkMlkYkchsnhubm7QaDSorKw03B6JqKuFBQSgr4eH4Y+Apo+72uuPPw53Z+dWl0tMS0OAjw/srK3veF2dpeG+6u791hBbxixcSkoKuyiJuom7uzsAsKuSup1MKoVEImn2WKXRoEalgkqtBgBodTrUqFSNugP1t/+v0WoNz2m02jYXJT7u7pA22LZer4dOr4der0ft7e0WlJbi//7730bb0DdZpum6jK2nrfR6PSprapo9X11bi4rq6kbPNdx3TfcjACgrK6HX67u0SGPLmIVLTk7G2LFjxY5B1Cu4ubkBAIqLi9G/f3+R0xDVtUb99csv8eysWYicNAnKykq8/Z//4N2nnkJqbi4+3rEDM8ePR3ZREU5cvoylv/89lBUVuHjjBvJKSvDZSy/B0c70BLNpeXlYt307Iu6+G4/eey9OxsVh3fbtWPDgg0jOysKZhAT8+5VXcCU5GYVKJX48eRIThw5FcnY2Nuzbhz9MmYItP/+MP82YgUH+/oZ1ebq4GF1PHxeXVt/z/86eRaFSCTcnJ/x65Qref/ppyKRSfLhtG343ZgyOXLyIvh4emHvffS2uRxAE/HvvXowYOBDfHzuGZZGR8Pfyau9H0CZsGbNwycnJCAoKEjsGUa/AljHqaYaHhGDq6NHILioCUDc4/eEJE+Ds4IDhISHQaLXwcHbGijlz8OTMmfju4EE8GhGBD5Ytg6ujIxLT01tcf4C3N5zt7Q2PJw4bBo1Oh34eHli1cCGmjxuH/WfPYvzQoQCA2RERGNC3L+4fMQLF5eXwdHXFt3/9K6aMGNFoXabW05q4lBRcuH4dix96CA+PHw8vNzfkl5Qgv6QEFdXVGDJgAKaPHYvDMTGtrutaRgbS8/IwIiQEL82b16XDfViMWTClUomioiJ2UxJ1E2dnZ8hkMhZj1KPMmjgRh2NioNJocDIuDhF33234mZVCAdfb4xu93dxgb2MDK3ldp5mbkxNKyspaXb9U2riUkMtkcHZwqFuHoyMyCgoMP5M0WMbGygr9vb1hZ2MDxe1tNlxXS+sx5dilSxjWoAFi5fz56NenD7zd3fH3J57AiStXcOnGDSgrK1tdV3C/fsgvKcGfP/4YyspK9PPwaPU1HcVizIJxWgui7iWVSuHi4sK5xqhHCerbFwE+Pjhy4QJ0ej2sFQqjy8mNtPx0+iipJuOxOptCLsf1jIxGzwmCgLLKSrz6xRcY5O+PUYMHt2ldcpkMn0VFITwwEC998gkudeEdD1iMWbDk5GTI5XKOXSHqRpyFn8TSdIB5w8ezJk7Elz/+iPHh4SaX0Wi10HdgkHr9QHxj66x/vr7Q0zSYw0sQBKDJ9hquy9h6WnP/iBE4dfUqTsbFQRAExFy7hoz8fBy9dAkqjQaerq4oKS9vtr6G26r//9nERFRUV2NZZCQWP/QQLt640aYMHcEB/Bbs5s2bCAgIgMLEX0FE1Pk48St1t+sZGbiemQm9IODeu+5CoVLZ6LGbkxMmDRuG45cvI8DHx/C6xLQ05JWU4FxSEgJ9fHD8yhVkFxYiLiUFLg4OSM7KggTAxKFD4dRgXFjTbd/MyoJao8Hk4cORkZ+PiupqnE1IgLebG2KuXUOBUonSigqMHzIEb/3nP3hyxgwUKJWoVaux5/RpLJo6FQ62to3W5ezgYHQ9eSUl8L59oYwxof7+iLj7brz97bdwdXDAkunTMWrQIAiCgE0HD+Kdb7/FhKFDodPrse3IEQwPCTHsK09X10b7DQDe37QJj0ZEoFatxvQuvBhOInB2Qov1xBNPIC8vDwcOHBA7ClGvMWPGDPTp0wfffvut2FHIgsydOxcAsGPHDgDA1X372n2jcK1OZ7Qrsjup1GpYW1l1+XZqVCoo5PJG71er0xmmrtDp9ZBJW+4crC+PKmtqjF5R6hkSgqEzZkAikSA6Ohr
```

Comment:
```suggestion
    "This pattern is more expensive to detect because it is implemented with a graph walk. Once the pass finds an incref, it considers each outgoing edge as a subtree and tries to find a decref at exit of the subtree. Once all the decref candidates are found, the pass walks backward along all incoming edges to find any violations. The backward walk must reach the incref node before reaching the function entry node. Also, no backward route can visit any decrefs."
```